source/server-discovery-and-monitoring/tests/rs/split_horizon_me_mismatch.yml

```diff
@@ -0,0 +1,60 @@
+description: Split horizon me mismatch
```

Comment:
Since it's a bug that split horizons works this way, I think we should leave split horizons out of the description/file name

There are currently three test files about mismatched me:

* primary_mismatched_me.yml
* secondary_mismatched_me.yml
* primary_to_no_primary_mismatched_me.yml

All of them have a mismatch between the host name in the seed list and in the "me" field, but no mismatch between the "me" field and the "hosts" field.  So how about we add three new test files that are just like those, except with a mismatch between "me" and "hosts" (but a match between the seed list and "hosts")

For clarity, we could rename the existing three files/descriptions to indicate what mismatch we're talking about.

Reply:
I'm not sure why we need three new test files, what does that cover that's not covered by this single one?

Reply:
For the same reason we have three test files for me mismatched with the seed list server name?  Are you implying that having three files for that is also unnecessary?  I'm just trying to keep things consistent.

Reply:
I'm just saying I don't think we gain meaningful coverage by adding those three new tests, not that the existing test files aren't valuable. My question is: do you see an active path here where we are likely to have additional bugs, or are you suggesting we triplicate just for consistency?

Reply:
I may be thinking about this wrong, so let's start over.

Since the server team agrees that the "me" field value should always match one of values in the hosts/passives/arbiters arrays, let's assume the server will have its own tests that assert that.  In that case, I don't think we need to add a new test case where that invariant is broken.

If we can agree on that (and I can see that we might not), is there a simpler test that keeps that invariant (and therefore is more similar to the existing three tests for mismatched me)  but still catches the current difference in behavior between Java/.NET and PyMongo/Node?

Reply:
Thanks. It turns out the position of the check in the pseudo code is wrong, and breaks the other me mismatch tests. I’ll fix that when I get in at noon. 

In the meantime, maybe I’m misunderstanding your initial ask. Would you be willing to provide an example test here? It would be interesting to see if it indeed expresses untested code paths for the SDAM update machinery

Reply:
> If we can agree on that (and I can see that we might not), is there a simpler test that keeps that invariant (and therefore is more similar to the existing three tests for mismatched me) but still catches the current difference in behavior between Java/.NET and PyMongo/Node?

Such a test (where the primary reports a "me" mismatch and the "me" is in the "hosts" field) already does exist. Actually two of them:

1. [primary_to_no_primary_mismatched_me.yml](https://github.com/mongodb/specifications/blob/0d7641c5560b11f195c0646d602d8c9e1f89f288/source/server-discovery-and-monitoring/tests/rs/primary_to_no_primary_mismatched_me.yml#L51)
2. [primary_mismatched_me.yml](https://github.com/mongodb/specifications/blob/0d7641c/source/server-discovery-and-monitoring/tests/rs/primary_mismatched_me.yml#L16)

Python actually does pass these tests even though we don't use the "me" mismatch rule on the primary's isMaster responses. We remove the primary because of another SDAM rule in [`updateRSFromPrimary`](https://github.com/mongodb/specifications/blob/0d7641c/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#updatersfromprimary):
```
for each server in topologyDescription.servers:
    if server.address not in description's "hosts", "passives", or "arbiters":
        remove the server and stop monitoring it
```

So there are two conclusions we can make:
- It's worthwhile to test the case where the hosts/me invariant is broken.
- Or the current test coverage is sufficient.

Reply:
Jeff and I talked this through today, and we think the behavior of the Node and Python driver is actually correct. I've replaced the test here with one validating the pseudocode provided in the SDAM specification.